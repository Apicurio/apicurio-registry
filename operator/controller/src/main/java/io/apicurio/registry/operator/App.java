package io.apicurio.registry.operator;

import io.fabric8.kubernetes.api.model.HasMetadata;
import io.fabric8.kubernetes.client.KubernetesClient;
import io.javaoperatorsdk.operator.Operator;
import io.javaoperatorsdk.operator.api.config.ConfigurationServiceOverrider;
import io.javaoperatorsdk.operator.api.reconciler.Reconciler;
import io.quarkus.runtime.Shutdown;
import io.quarkus.runtime.Startup;
import jakarta.annotation.Priority;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.enterprise.inject.Instance;
import jakarta.enterprise.inject.Produces;
import jakarta.inject.Inject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Arrays;
import java.util.function.Consumer;
import java.util.stream.Collectors;

import static io.apicurio.registry.operator.utils.Utils.isBlank;
import static io.quarkus.runtime.configuration.ConfigUtils.isProfileActive;
import static org.eclipse.microprofile.config.ConfigProvider.getConfig;

@ApplicationScoped
public class App {

    private static final Logger log = LoggerFactory.getLogger(App.class);

    @Inject
    Instance<Reconciler<? extends HasMetadata>> reconcilers;

    @Inject
    KubernetesClient client;

    private Operator operator;

    @Startup
    void startup() {
        if (isProfileActive("test")) {
            log.info("Operator is not started automatically during testing.");
        } else {
            start();
        }
    }

    public void start() {
        start(configOverride -> {
            configOverride.withKubernetesClient(client);
            configOverride.withUseSSAToPatchPrimaryResource(false);
        });
    }

    public void start(Consumer<ConfigurationServiceOverrider> configOverride) {
        log.info("Starting the Apicurio Registry 3 Operator version {}", getConfig().getValue("registry.version", String.class));
        operator = new Operator(configOverride);

        var watchedNamespacesRaw = getConfig().getOptionalValue("apicurio.operator.watched-namespaces", String.class).orElse("");
        log.debug("apicurio.operator.watched-namespaces={}", watchedNamespacesRaw);

        reconcilers.forEach(r -> {
            operator.register(r, controllerConfigOverride -> {
                if (isBlank(watchedNamespacesRaw)) {
                    controllerConfigOverride.watchingAllNamespaces();
                    log.info("Watching all namespaces.");
                } else {
                    var watchedNamespaces = Arrays.stream(watchedNamespacesRaw.split(",")).map(String::trim).collect(Collectors.toSet());
                    controllerConfigOverride.settingNamespaces(watchedNamespaces);
                    log.info("Watching namespace(s): {}", watchedNamespaces.stream().map(n -> "'" + n + "'").collect(Collectors.joining(", ")));
                }
            });
        });
        operator.start();
        log.info("Operator started");
    }

    @Shutdown
    void shutdown() {
        if (isProfileActive("test")) {
            log.info("Operator is not stopped automatically during testing.");
        } else {
            stop();
        }
    }

    public void stop() {
        log.info("Stopping the Apicurio Registry 3 Operator.");
        operator.stop();
    }

    @Produces
    @Priority(100)
    public Operator produce() {
        return operator;
    }
}
