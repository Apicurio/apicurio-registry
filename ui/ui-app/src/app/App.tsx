import "./App.css";
import "@patternfly/patternfly/patternfly.css";
import "@patternfly/patternfly/patternfly-addons.css";

import { FunctionComponent } from "react";
import { Page } from "@patternfly/react-core";
import { BrowserRouter as Router, Route, Routes } from "react-router-dom";
import { AppHeader, ApplicationAuth } from "@app/components";
import {
    ArtifactRedirectPage,
    ArtifactsPage,
    ArtifactVersionPage,
    NotFoundPage,
    RootRedirectPage,
    RulesPage
} from "@app/pages";
import { Services } from "../services";
import { RolesPage, SettingsPage } from "./pages";

export type AppProps = {
    // No props
};

/**
 * The main application class.
 */
export const App: FunctionComponent<AppProps> = () => {
    const contextPath: string | undefined = Services.getConfigService().uiContextPath();
    Services.getLoggerService().info("[App] Using app contextPath: ", contextPath);

    return (
        <ApplicationAuth>
            <Router basename={contextPath}>
                <Page
                    className="pf-m-redhat-font"
                    isManagedSidebar={false}
                    header={<AppHeader />}
                >
                    <Routes>
                        <Route path="/" element={ <RootRedirectPage /> } />
                        <Route path="/rules" element={ <RulesPage /> } />
                        <Route path="/roles" element={ <RolesPage /> } />
                        <Route path="/settings" element={ <SettingsPage /> } />
                        <Route path="/artifacts" element={ <ArtifactsPage /> } />
                        <Route
                            path="/artifacts/:groupId/:artifactId"
                            element={ <ArtifactRedirectPage /> }
                        />
                        <Route
                            path="/artifacts/:groupId/:artifactId/versions/:version"
                            element={ <ArtifactVersionPage /> }
                        />
                        <Route element={ <NotFoundPage /> } />
                    </Routes>
                </Page>
            </Router>
        </ApplicationAuth>
    );
};
