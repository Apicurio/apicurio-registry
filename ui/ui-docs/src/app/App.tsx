import "./App.css";
import { FunctionComponent, useEffect, useState } from "react";
import { useWindow } from "@app/hooks/useWindow.ts";
import { RedocStandalone } from "redoc";
import { DEMO_OPENAPI } from "@app/DemoData.ts";
import { If } from "@app/components/common/If.tsx";

enum ContentType {
    OPENAPI = "OPENAPI", OTHER = "OTHER"
}

export type DocsUpdateType = {
    content: any;
    contentType: ContentType;
};

export type AppProps = {
    // No props
};

export const App: FunctionComponent<AppProps> = () => {
    const [content, setContent] = useState<any>({});
    const [contentType, setContentType] = useState(ContentType.OTHER);

    const w: Window = useWindow();
    const queryParams = new URLSearchParams(w.location.search);

    useEffect(() => {
        const isDemo: boolean = queryParams.get("demo") === "true";
        if (isDemo) {
            setContent(DEMO_OPENAPI);
            setContentType(ContentType.OPENAPI);
        } else {
            w.onmessage = (evt: MessageEvent) => {
                console.info("[App] OnMessage received: ", evt);
                if (evt.data && evt.data.type && evt.data.type === "apicurio-docs-render") {
                    const update: DocsUpdateType = evt.data.data as DocsUpdateType;
                    console.info("[App] Render docs message received from parent frame: ", update.contentType);
                    setContent(update.content);
                    setContentType(update.contentType);
                }
            };
        }
    }, []);

    return (
        <>
            <If condition={contentType === ContentType.OPENAPI}>
                <RedocStandalone spec={content} />
            </If>
        </>
    );
};
