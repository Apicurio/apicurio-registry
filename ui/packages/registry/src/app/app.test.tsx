import React from 'react';
import App from './app';
import { mount, shallow } from 'enzyme';
import { Alert, Button } from '@patternfly/react-core';

describe('App tests', () => {
  test('should render default App component', () => {
    const view = shallow(<App />);
    expect(view).toMatchSnapshot();
  });

  it('should render a dismiss button', () => {
    const wrapper = mount(<App />);
    const button = wrapper.find(Button);
    expect(button.exists()).toBe(true)
  });

  it('should hide the alert when clicking the dismiss button', () => {
    const wrapper = mount(<App />);
    const button = wrapper.find(Button);
    expect(wrapper.find(Alert).exists()).toBe(true);
    button.simulate('click');
    expect(wrapper.find(Alert).exists()).toBe(false);
  });
});
