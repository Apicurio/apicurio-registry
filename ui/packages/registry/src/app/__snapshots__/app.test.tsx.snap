// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests should render default App component 1`] = `
<Fragment>
  <BackgroundImage
    className=""
    src={
      Object {
        "filter": "test-file-stub#image_overlay",
        "lg": "test-file-stub",
        "sm": "test-file-stub",
        "sm2x": "test-file-stub",
        "xs": "test-file-stub",
        "xs2x": "test-file-stub",
      }
    }
  />
  <div
    className="app-container"
  >
    <div
      className="notification-container"
    >
      <Alert
        action={
          <AlertActionCloseButton
            aria-label=""
            className=""
            onClose={[Function]}
          />
        }
        className=""
        title="Setup Complete"
        variant="success"
        variantLabel={null}
      >
        You have successfully launched your patternfly starter project.
      </Alert>
    </div>
  </div>
</Fragment>
`;
